### ISSUE TITLE
Dark mode not working because of `@layer base`, demo repo provided

### ISSUE THREAD

**Issue by Zwyx** [shadcn-ui/ui#313]
Hi and congrats for making these beautiful components!

I notice that dark mode isn't working on a fresh install:

- ran `npx create-next-app@latest`, didn't choose to use the `app` directory
- ran `npx shadcn-ui init`
- ran `npm i next-themes`
- added `<ThemeProvider attribute="class">` in `_app.tsx`

The result: we notice that the `dark` class is present (thanks to `prefers-color-scheme: dark`), but it's the variables from `:root` that are applied, not the ones from `.dark`:

![image](https://user-images.githubusercontent.com/29386932/236669470-74bd8822-28f4-48d6-987e-087276213963.png)

Demo repository:
https://github.com/Zwyx/shadcn-ui-dark-mode-demo deployed at https://zwyx.github.io/shadcn-ui-dark-mode-demo/

Notes:
- Removing `@layer base` from `globals.css` fixes this.
- This issue is very similar to https://github.com/shadcn/ui/issues/278 although I'm not using the `app` directory, and I'm not able to fix it...

**Comment by luanfvieira** [comment]
Update your globals.css

```css
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {
  :root {
    --background: 0 0% 100%;
    --foreground: 222.2 47.4% 11.2%;

    --muted: 210 40% 96.1%;
    --muted-foreground: 215.4 16.3% 46.9%;

    --popover: 0 0% 100%;
    --popover-foreground: 222.2 47.4% 11.2%;

    --border: 214.3 31.8% 91.4%;
    --input: 214.3 31.8% 91.4%;

    --card: 0 0% 100%;
    --card-foreground: 222.2 47.4% 11.2%;

    --primary: 222.2 47.4% 11.2%;
    --primary-foreground: 210 40% 98%;

    --secondary: 210 40% 96.1%;
    --secondary-foreground: 222.2 47.4% 11.2%;

    --accent: 210 40% 96.1%;
    --accent-foreground: 222.2 47.4% 11.2%;

    --destructive: 0 100% 50%;
    --destructive-foreground: 210 40% 98%;

    --ring: 215 20.2% 65.1%;

    --radius: 0.5rem;
  }

  .dark {
    --background: 224 71% 4%;
    --foreground: 213 31% 91%;

    --muted: 223 47% 11%;
    --muted-foreground: 215.4 16.3% 56.9%;

    --accent: 216 34% 17%;
    --accent-foreground: 210 40% 98%;

    --popover: 224 71% 4%;
    --popover-foreground: 215 20.2% 65.1%;

    --border: 216 34% 17%;
    --input: 216 34% 17%;

    --card: 224 71% 4%;
    --card-foreground: 213 31% 91%;

    --primary: 210 40% 98%;
    --primary-foreground: 222.2 47.4% 1.2%;

    --secondary: 222.2 47.4% 11.2%;
    --secondary-foreground: 210 40% 98%;

    --destructive: 0 63% 31%;
    --destructive-foreground: 210 40% 98%;

    --ring: 216 34% 17%;

    --radius: 0.5rem;
  }
}

@layer base {
  * {
    @apply border-border;
  }
  body {
    @apply bg-background text-foreground;
    font-feature-settings: "rlig" 1, "calt" 1;
  }
}

@layer utilities {
  .step {
    counter-increment: step;
  }

  .step:before {
    @apply absolute w-8 h-8 bg-muted rounded-full font-medium text-center text-base inline-flex items-center justify-center -indent-px;
    @apply ml-[-41px];
    content: counter(step);
  }
}

@media (max-width: 640px) {
  .container {
    @apply px-4;
  }
}

**Comment by Zwyx** [comment]
Would you be able to check out this repo: https://zwyx.github.io/shadcn-ui-dark-mode-demo/

Then replace the content of `globals.css` by yours, start the app, open it in an incognito window (to make sure there is not "theme" value in local storage), open the dev tools and turn on `prefers-color-scheme: dark` like it is shown on the screenshot above.

I did exactly that, and the dark theme still not apply. Even with the content of your file.

**Comment by hgaron-blockbar** [comment]
@Zwyx can you please check your `tailwind.config.js` file? I had to update mine because I'm using a `src` directory - so had to account for that.

example:
`content: ["src/app/**/*.{ts,tsx}", "components/**/*.{ts,tsx}"],`

By default, there's no `src` prefix there (I realize you aren't using `app` dir, but maybe the path in the `content` directory needs to be corrected?

**Comment by Zwyx** [comment]
The `tailwind.config.js` file, present here https://github.com/Zwyx/shadcn-ui-dark-mode-demo/blob/main/tailwind.config.js seems correctly configured.

**Comment by hgaron-blockbar** [comment]
@Zwyx I haven't yet gotten to the root of the issue, but it does seem to be some sort of config issue. Updating this as the `content` array restores the styles as expected:

`content: ["**/*.{ts,tsx}"]`

Not certain yet why the original config isn't working.

**Comment by nekochan0122** [comment]
I got the same problem

**Comment by shadcn** [comment]
@Zwyx Is this still not working? How can I help?

**Comment by bidodev** [comment]
Got the same problem, following the previous suggestions and replacing the content in the the tailwind config to the following 
`content to ["**/*.{ts,tsx}"]` solve the issue.

**Comment by Zwyx** [comment]
Hi @shadcn, thanks for your message. Yes, this is still an issue (I've just retried from scratch by redoing the steps described in the first post; same result).

I'm not familiar with the `@layer` CSS tag, and I've noticed that if I remove it from `globals.css`, all works correctly.

I don't know if you can help and I definitely don't want to waste your time, but I'd be interesting to know **why you use `@layer`?** And if you can think of any issue that could appear by removing `@layer` like I did.

The user hgaron-blockbar, above, noticed that we can also make it work by tweaking Tailwind's config.

These two ways of solving the issue are workarounds, but don't really identify the root of the problem.

**Comment by ndinata** [comment]
I had the same problem until I found out I had to add `themeColor` to my root layout metadata, following the [template repo here](https://github.com/shadcn/next-template/blob/9d67d5bbf68b3b5bfcf69ea45c641b5936616313/app/layout.tsx#L17). No other special changes to `tailwind.config.js` nor `globals.css`.

@Zwyx I notice you're not using the `app` dir in your demo repo, but since `themeColor` gets converted into `<meta name="theme-color" />` tags (according to [Nextjs docs](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#themecolor)), maybe you could try copy-pasting the appropriate tags into your project and see if dark mode works afterwards?

**Comment by Zwyx** [comment]
Hi Nico and thanks for your message. Good find! That indeed fixed it. I simply added the following in `_documents.tsx`:

```tsx
	<Head>
		<meta
			name="theme-color"
			media="(prefers-color-scheme: light)"
			content="white"
		/>
		<meta
			name="theme-color"
			media="(prefers-color-scheme: dark)"
			content="black"
		/>
	</Head>
}

```

I don't see the relation between `@layer base` and `theme-color` though. It could be interesting to dig, but I'm not keen to dive in a deep rabbit hole.

I also noticed something else: this problem was happening in Chrome, but not in Firefox.

I guess with this third workaround, we could close the issue if you'd like, Shadcn?

**Comment by chen-rn** [comment]
@ndinata worked for me too!

@shadcn We should either add this to the docs or somehow fix it!

**Comment by yannxaver** [comment]
Yes, please add this to the docs. I was just pulling out my hair why dark mode was not working because the setup is pretty simple using `next-themes`. Can't be the only one 😁

**Comment by Zwyx** [comment]
Made a PR for this: https://github.com/shadcn/ui/pull/938

**Comment by CibiAananth** [comment]
Adding this in `app/layout.tsx` worked for me

```
import { Metadata } from 'next';

export const metadata: Metadata = {
  themeColor: [
    { media: '(prefers-color-scheme: light)', color: 'white' },
    { media: '(prefers-color-scheme: dark)', color: 'black' },
  ],
};
```

**Comment by marcklingen** [comment]
Had the same issue, fixed it with `:root[class~="dark"]` instead of `.dark` in `style.css`

style.css for reference: https://github.com/langfuse/langfuse-docs/blob/e9ba748e0e53e5d9179ffc0040839108d1930621/style.css#L39

**Comment by cleytoncarvalho** [comment]
> Had the same issue, fixed it with `:root[class~="dark"]` instead of `.dark` in `style.css`
> 
> style.css for reference: https://github.com/langfuse/langfuse-docs/blob/e9ba748e0e53e5d9179ffc0040839108d1930621/style.css#L39

Thank you! Worked just fine. I was searching for a solution for hours.

**Comment by benyamynbrkyc** [comment]
> Had the same issue, fixed it with `:root[class~="dark"]` instead of `.dark` in `style.css`
> 
> style.css for reference: https://github.com/langfuse/langfuse-docs/blob/e9ba748e0e53e5d9179ffc0040839108d1930621/style.css#L39

Thank you, this worked for me as well. I just set the styles up with both targets, class and root:

```css
  .dark,
  :root[class~="dark"] {
    ...
  }
```

**Comment by toitzi** [comment]
Can confirm @benyamynbrkyc solution worked for me as well, imho a good solution since you dont need to modify your tailwind config or meta data tags.

**Comment by alabenyahia** [comment]
For me, on a Vite/React app, the problem was forgetting to import the global.css file into my App.tsx.

**Comment by mortenjust** [comment]
I seem to have fixed this by adding “dark” to the tailwind safe list

**Comment by xieyezi** [comment]
I've also encountered the same issue. Here's my setup:

```
"class-variance-authority": "^0.7.0",
"clsx": "^2.0.0",
"lucide-vue-next": "^0.292.0",
"tailwind-merge": "^2.0.0",
"tailwindcss-animate": "^1.0.7",
```
and my `components.json` :

```json

{
  "style": "default",
  "typescript": true,
  "tailwind": {
    "config": "tailwind.config.js",
    "css": "assets/css/tailwind.css",
    "baseColor": "zinc",
    "cssVariables": true
  },
  "framework": "nuxt",
  "aliases": {
    "components": "@/components",
    "utils": "@/lib/utils"
  }
}
```

When I add `class="dark"`  to the HTML tag, my div is set as follows:

```html

  <div class="bg-background text-foregroundl">
   ...
  </div>
```
However, this doesn't have any effect, and the page still appears in light mode.

**Comment by dong20191111** [comment]
这个问题确实困扰了我好几个小时，直到我仔细的一遍一遍看shadcn的文档才解决，我是用shadcn+tailwindcss，因为shadcn的代码可以直接生成，所以用shadcn的provider方式处理。但是怎么也不好用。这个问题应该与我的是一个问题。
1、按照https://ui.shadcn.com/docs/dark-mode/vite先增加provider
2、页面点击切换主题，会发现<html lang="en" class="dark" foxified="">，其中class="dark"是可以变化为light的，说明provider是可用的
3、我们通过https://ui.shadcn.com/themes，定制自己的主题，拿到代码并放到global.css中，这个样式文件与components.json里的tailwind样式文件一致：
"tailwind": {
    "config": "tailwind.config.js",
    "css": "src/globals.css",
    "baseColor": "gray",
    "cssVariables": true
  },
4、仔细阅读https://ui.shadcn.com/docs/theming，有两种控制主题样式的方式：
 一种是dark:bg-white这样的，在tailwind里边直接用的这种。另一种是变量形式：
<div className="bg-background text-foreground" />
其中的background，foreground并不是tailwind自带的样式变量，需要我们要定义才能识别。
注意"cssVariables": true，要设置为true才可以用这种方式。因为我们要用的主题样式控制都是这种的：
@layer base {
  :root {
    --background: 0 0% 100%;
    --foreground: 20 14.3% 4.1%;
    --card: 0 0% 100%;
    --card-foreground: 20 14.3% 4.1%;
    --popover: 0 0% 100%;
5、在tailwind.config.js文件中增加自定义颜色：
colors: {
        'background':"hsl(var(--background))",
        "foreground": "hsl(var(--foreground))",
![image](https://github.com/shadcn-ui/ui/assets/48936035/908294d4-0953-48ee-ba5b-6c4c5e2910c3)
文档这里其实已经举例了一种增加主题控制的warning颜色的处理方法，就是没直接写整个页面的背景及前景色怎么做。。。
也许不想让我们直接copy用吧，还是想让我们动动脑筋。
6、最后光上边这些还是没有效果的，因为还要给body加上class这样才能找到对应的点去处理主题颜色
我们在最顶端的body增加class:
<body class="bg-background text-foreground">
这样就全部完成了。

**Comment by shadcn** [comment]
Is this still an issue?

**Comment by marcklingen** [comment]
> Had the same issue, fixed it with `:root[class~="dark"]` instead of `.dark` in `style.css`
> 
> 
> 
> style.css for reference: https://github.com/langfuse/langfuse-docs/blob/e9ba748e0e53e5d9179ffc0040839108d1930621/style.css#L39

@shadcn this works well for us

**Comment by benyamynbrkyc** [comment]
Looks like it is. I still have to use the root selector, e.g.:

```css
.dark,
  :root[class~='dark'] {
      ...
}
```

**Comment by Zwyx** [comment]
There is a PR for that: https://github.com/shadcn-ui/ui/pull/1780

**Comment by AchrafGarai** [comment]
This code fixed it for me 
`@tailwind base;
@tailwind components;
@tailwind utilities;


:root {
    --background: 0 0% 100%;
    --foreground: 222.2 47.4% 11.2%;
 
    --muted: 210 40% 96.1%;
    --muted-foreground: 215.4 16.3% 46.9%;
 
    --popover: 0 0% 100%;
    --popover-foreground: 222.2 47.4% 11.2%;
 
    --card: 0 0% 100%;
    --card-foreground: 222.2 47.4% 11.2%;
 
    --border: 214.3 31.8% 91.4%;
    --input: 214.3 31.8% 91.4%;
 
    --primary: 222.2 47.4% 11.2%;
    --primary-foreground: 210 40% 98%;
 
    --secondary: 210 40% 96.1%;
    --secondary-foreground: 222.2 47.4% 11.2%;
 
    --accent: 210 40% 96.1%;
    --accent-foreground: 222.2 47.4% 11.2%;
 
    --destructive: 0 100% 50%;
    --destructive-foreground: 210 40% 98%;
 
    --ring: 215 20.2% 65.1%;
 
    --radius: 0.5rem;
}

@media (prefers-color-scheme: dark) {
  :root {
    --background: 224 71% 4%;
    --foreground: 213 31% 91%;
 
    --muted: 223 47% 11%;
    --muted-foreground: 215.4 16.3% 56.9%;
 
    --popover: 224 71% 4%;
    --popover-foreground: 215 20.2% 65.1%;
 
    --card: 224 71% 4%;
    --card-foreground: 213 31% 91%;
 
    --border: 216 34% 17%;
    --input: 216 34% 17%;
 
    --primary: 210 40% 98%;
    --primary-foreground: 222.2 47.4% 1.2%;
 
    --secondary: 222.2 47.4% 11.2%;
    --secondary-foreground: 210 40% 98%;
 
    --accent: 216 34% 17%;
    --accent-foreground: 210 40% 98%;
 
    --destructive: 0 63% 31%;
    --destructive-foreground: 210 40% 98%;
 
    --ring: 216 34% 17%;
 
    --radius: 0.5rem;
  }
}

  * {
    @apply border-border;
  }
  body {
    @apply bg-background text-foreground;
    font-feature-settings: "rlig" 1, "calt" 1;
  }`

**Comment by AchrafGarai** [comment]
this fixed it for me
```
@tailwind base;
@tailwind components;
@tailwind utilities;


:root {
    --background: 0 0% 100%;
    --foreground: 222.2 47.4% 11.2%;
 
    --muted: 210 40% 96.1%;
    --muted-foreground: 215.4 16.3% 46.9%;
 
    --popover: 0 0% 100%;
    --popover-foreground: 222.2 47.4% 11.2%;
 
    --card: 0 0% 100%;
    --card-foreground: 222.2 47.4% 11.2%;
 
    --border: 214.3 31.8% 91.4%;
    --input: 214.3 31.8% 91.4%;
 
    --primary: 222.2 47.4% 11.2%;
    --primary-foreground: 210 40% 98%;
 
    --secondary: 210 40% 96.1%;
    --secondary-foreground: 222.2 47.4% 11.2%;
 
    --accent: 210 40% 96.1%;
    --accent-foreground: 222.2 47.4% 11.2%;
 
    --destructive: 0 100% 50%;
    --destructive-foreground: 210 40% 98%;
 
    --ring: 215 20.2% 65.1%;
 
    --radius: 0.5rem;
}

@media (prefers-color-scheme: dark) {
  :root {
    --background: 224 71% 4%;
    --foreground: 213 31% 91%;
 
    --muted: 223 47% 11%;
    --muted-foreground: 215.4 16.3% 56.9%;
 
    --popover: 224 71% 4%;
    --popover-foreground: 215 20.2% 65.1%;
 
    --card: 224 71% 4%;
    --card-foreground: 213 31% 91%;
 
    --border: 216 34% 17%;
    --input: 216 34% 17%;
 
    --primary: 210 40% 98%;
    --primary-foreground: 222.2 47.4% 1.2%;
 
    --secondary: 222.2 47.4% 11.2%;
    --secondary-foreground: 210 40% 98%;
 
    --accent: 216 34% 17%;
    --accent-foreground: 210 40% 98%;
 
    --destructive: 0 63% 31%;
    --destructive-foreground: 210 40% 98%;
 
    --ring: 216 34% 17%;
 
    --radius: 0.5rem;
  }
}

  * {
    @apply border-border;
  }
  body {
    @apply bg-background text-foreground;
    font-feature-settings: "rlig" 1, "calt" 1;
  }
```

**Comment by Zwyx** [comment]
@AchrafGarai you won't be able to allow your users to set the them they want; dark theme will only be applied if it's their OS's preference. Which isn't a problem, as long as it's what you want.

**Comment by rafaelcalpena** [comment]
I was also trying to enable dark theme with `shadcn-ui` yesterday and ended up on this thread after searching for a fix. Here are some observations and possible solutions:

**TL;DR:** 
- Add [`safelist: ['dark']`](https://tailwindcss.com/docs/content-configuration#safelisting-classes) to your `tailwind.config.ts` if you don't mind changing such file and you want an officially supported fix;
**OR**
- Replace the `.dark` selector in your [`globals.css` file](https://github.com/shadcn-ui/ui/issues/313#issuecomment-1676513456) file with `:root[class~="dark”]` (thanks @marcklingen).

## Explanation

- [This page](https://tailwindcss.com/docs/content-configuration) states that Tailwind scans all files matched by your `content` glob patterns (e.g. `["**/*.{ts,tsx}”]` will scan for all files ending with `.ts` or `.tsx` relative to your project root). 
- [Their documentation](https://tailwindcss.com/docs/content-configuration#working-with-third-party-libraries) suggests that libraries should write styles *without* using Tailwind’s `@layer` feature, in order to ensure that `“Tailwind always includes those styles in your CSS”`. Additionally,

> The way Tailwind scans your source code for classes is intentionally very simple — we don’t actually parse or execute any of your code in the language it’s written in, we just use regular expressions to extract every string that could possibly be a class name.

Therefore, it seems that the main cause for the dark theme not being applied is that **Tailwind has not found any project files containing a `dark` class call, thus removing it from the final bundle**. 

## Workarounds and Fixes

Using Storybook and Next.js, the following fixes have worked for me: 

> **NOTE:** Depending on your setup, Tailwind may cache styles even after removed, so I recommend to **restart your server** after applying/removing any of the following fixes/hacks.

- Defining the `.dark` class outside of the `@layer` in your `globals.css` file will always include dark theme styles in the final bundle, but it might break styling since the original code written by @shadcn could rely on the order of `@layer`s, affecting the priority of class resolution.

- Add any object containing a string with `dark` in a file scanned by Tailwind. **Not recommended**, since this is a hack and it’s hard to locate and debug:

   `const triggerTailwind = ['dark'];`

- The above may be the reason why many different solutions in this thread work:

    - Using [`content: ["**/*.{ts,tsx}”]`](https://github.com/shadcn-ui/ui/issues/313#issuecomment-1550402125) (thanks @hgaron-blockbar)  works because there must be a string containing `dark` somewhere in your code. However, Tailwind in general recommends [against using broad patterns](https://tailwindcss.com/docs/content-configuration#pattern-recommendations).

    - [This solution which adds a `themeColor` object](https://github.com/shadcn-ui/ui/issues/313#issuecomment-1557160587) (thanks @ndinata and @Zwyx) is probably triggering the `dark` class compilation because it contains `dark` inside `themeColor.media`:
      ``` 
      themeColor: [
          { media: "(prefers-color-scheme: dark)", color: "black" },
      ]
      ```

- Adding the global css file `./src/app/globals.css` to your `content` property in `tailwind.config.ts` works as Tailwind will probably scan it and find the `.dark` class, however Tailwind also explicitly recommends [against adding `.css` files to your `content` folder](https://tailwindcss.com/docs/content-configuration#pattern-recommendations):

    > It’s also important that you don’t scan any CSS files — configure Tailwind to scan your templates where your class names are being used, never the CSS file that Tailwind is generating.

- I really liked @marcklingen’s solution: by replacing `.dark` with `:root[class~="dark"]` in the `globals.css` file, it seems that Tailwind identifies the `"dark"` token as a string, or maybe it has a special rule for such selectors inside `@layer`s. This  will continue to work unless Tailwind changes its scanning system in a way that affects the special `:root[class~="dark"]` selector.

- If you **don’t mind changing your `tailwind.config.ts` file**, these strategies are also valid:
  
  - Adding `{ raw: '<div class=“dark”>’, extension: 'html' }` as [Raw Content](https://tailwindcss.com/docs/content-configuration#configuring-raw-content) inside your `content` property works, but it is hacky and there is a better way below:

  - **Recommended:** add the [safelist property](https://tailwindcss.com/docs/content-configuration#safelisting-classes) and provide the `dark` class to it: `safelist: ["dark"]`. 


@shadcn Thank you for creating such an awesome collection of components!

**Comment by Zwyx** [comment]
Thank you so much @rafaelcalpena for this amazing explanation!

You mentioned « _the original code written by @shadcn could rely on the order of `@layer`s_ ». Do you know more about this? This is something I've been very keen to understand since this issue started: **are these `@layer`s useful?!**

Because simply removing them solves this issue completely. So it's what's I've done in my 5-10 projects that use Shadcn UI, and it never caused me any issues.

**Comment by rafaelcalpena** [comment]
> Thank you so much @rafaelcalpena for this amazing explanation!
> 
> You mentioned « _the original code written by @shadcn could rely on the order of `@layer`s_ ». Do you know more about this? This is something I've been very keen to understand since this issue started: **are these `@layer`s useful?!**
> 
> Because simply removing them solves this issue completely. So it's what I've done in my 5-10 projects that use Shadcn UI, and it never caused me any issues.

Thank you @Zwyx. Yes, the `layers` are useful to dictate the priority order for the styles to be applied. You can think of them as a way to augment the default behavior in CSS, where the "last rule declaration wins" (explained [here](https://tailwindcss.com/docs/adding-custom-styles#using-css-and-layer)). The difference is that with `@layer`s you can write rules in arbitrary order and Tailwind will "move" them to the appropriate precedence layer during compilation.

By default Tailwind has 3 layers: `base`, `components` and `utilities`:
```
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {
    .my-class {background: red;}
}

```
In case of conflicts, the last layer should win, so we have `utilities > components > base (with .my-class inside)`. 
By taking `.my-class` away from the `@layer base`, I assume the order would become  `.my-class > utilities > components > base`, thus changing the precedence.

Some links that explain layers more in-depth: [Organize your CSS in the Tailwind style with @layer directive](https://bloggie.io/@kinopyo/organize-your-css-in-the-tailwind-style-with-layer-directive) and [Tailwind's Official Documentation](https://tailwindcss.com/docs/adding-custom-styles#using-css-and-layer)

On the other hand, it is great to hear that moving `.dark` outside of `base` has worked well and without conflicts, so it is also a useful fix.

**Comment by Zwyx** [comment]
Excellent. It makes much more sense now. I didn't realise Tailwind uses `@layer`. Thanks a lot Rafael!

I think we now have everything to take a decision on this issue:

**TL;DR:**

- reason the problem is not experienced by everyone:
  - **it only happens when `dark:` isn't used anywhere**, which makes Tailwind dropping it from the CSS.

- **fix**:
**EITHER**
  - **replace `.dark` by `:root[class~="dark"]` in `globals.css`** – there is an [old PR](https://github.com/shadcn-ui/ui/pull/1780) for that, I'd be happy to update if it's the solution that is chosen.
**OR**
  - **add `safelist: ['dark']` to `tailwind.config.ts`**, which sounds a bit easier.

**@shadcn, which one of these two solutions would you like to see implemented?**

Thanks again heaps to @rafaelcalpena for his massive help!

**Comment by thomasweigert91** [comment]
For me, replacing .dark with :root[data-theme='dark'] worked!

**Comment by Maniktherana** [comment]
I'm having an issue where dark mode works in chrome but not safari 
I've tried everything mentioned here, but nothing seems to work

**Comment by FrancescoSaverioZuppichini** [comment]
Hi there, I am experiencing something that I think is related, by following the example for [dark mode](https://ui.shadcn.com/docs/dark-mode/next) my toggle has the two icons one over the other, this is because `dark` doesn't get applied I fear

![Screenshot 2024-04-22 at 22 54 20](https://github.com/shadcn-ui/ui/assets/15908060/8e1049a0-4d5a-4a5f-803d-0873bd0aab77)

I've tried the options provided by @Zwyx but neither has worked.

I've just installed it, linking my project here

https://github.com/FrancescoSaverioZuppichini/nextjs-cloudfare/tree/dev

Thanks a lot

**Comment by ShaunLWM** [comment]
Not sure if it's related as well. Plain vite + react-ts template + vite shadcn.
```
  "dependencies": {
    "@radix-ui/react-dropdown-menu": "^2.0.6",
    "@radix-ui/react-slot": "^1.0.2",
    "class-variance-authority": "^0.7.0",
    "clsx": "^2.1.1",
    "lucide-react": "^0.378.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "tailwind-merge": "^2.3.0",
    "tailwindcss-animate": "^1.0.7"
  },
```
https://github.com/shadcn-ui/ui/assets/37434846/8bc75298-9f64-420e-863c-4bdf388eff51

**Comment by Enstso** [comment]
> Not sure if it's related as well. Plain vite + react-ts template + vite shadcn.
> 
> ```
>   "dependencies": {
>     "@radix-ui/react-dropdown-menu": "^2.0.6",
>     "@radix-ui/react-slot": "^1.0.2",
>     "class-variance-authority": "^0.7.0",
>     "clsx": "^2.1.1",
>     "lucide-react": "^0.378.0",
>     "react": "^18.2.0",
>     "react-dom": "^18.2.0",
>     "tailwind-merge": "^2.3.0",
>     "tailwindcss-animate": "^1.0.7"
>   },
> ```
> 
>  Screen.Recording.2024-05-17.at.09.35.57.mov

Hello do you fix it ? i have a same problem

**Comment by Nawabul** [comment]
**Check**: Any another tag does not have dark or light class .
if there is any tag have dark or light class then there children will act relatively.

**Comment by shadcn** [comment]
This issue has been automatically closed because it received no activity for a while. If you think it was closed by accident, please leave a comment. Thank you.

**Comment by Zwyx** [comment]
Anyone experiencing this issue, please refer to [this comment](https://github.com/shadcn-ui/ui/issues/313#issuecomment-1929054475), which contains the reason, a way to fix it, and a link to a PR.

**Comment by rhyek** [comment]
Not sure if anyone else is seeing this, but for a new Next 14 app with app router + shadcn, I was testing a few of the dashboard blocks and with the default "light" theme everything looked off. Like it was a mix of dark and light mode. Looking at the dark theme (manually added the `dark` class to `<body>`) it looked as it should.

I changed the `@media (prefers-color-scheme: dark) {` line to `:root[class~="dark"] {` and now both light and dark themes look right.

I'm on macos and I've set it's Appearance to "Dark". That probably has something to do with it.

With this change the user will have to change themes manually, which I think is ok.

**Comment by alxgrgrs** [comment]
> Not sure if anyone else is seeing this, but for a new Next 14 app with app router + shadcn, I was testing a few of the dashboard blocks and with the default "light" theme everything looked off. Like it was a mix of dark and light mode. Looking at the dark theme (manually added the `dark` class to `<body>`) it looked as it should.
> 
> I changed the `@media (prefers-color-scheme: dark) {` line to `:root[class~="dark"] {` and now both light and dark themes look right.
> 
> I'm on macos and I've set it's Appearance to "Dark". That probably has something to do with it.
> 
> With this change the user will have to change themes manually, which I think is ok.

This does the fix. Not sure why this isn't in the documentation.

**Comment by Songkeys** [comment]
This puzzled me for hours! Thank you so much https://github.com/shadcn-ui/ui/issues/313#issuecomment-1929054475

Please. Why don't we merge that PR?

**Comment by chrispetrou** [comment]
> Not sure if anyone else is seeing this, but for a new Next 14 app with app router + shadcn, I was testing a few of the dashboard blocks and with the default "light" theme everything looked off. Like it was a mix of dark and light mode. Looking at the dark theme (manually added the `dark` class to `<body>`) it looked as it should.
> 
> I changed the `@media (prefers-color-scheme: dark) {` line to `:root[class~="dark"] {` and now both light and dark themes look right.
> 
> I'm on macos and I've set its Appearance to "Dark". That probably has something to do with it.
> 
> With this change the user will have to change themes manually, which I think is ok.

I had the exact same issue. I did modify `globals.css` as per [this comment](https://github.com/shadcn-ui/ui/issues/313#issuecomment-1927525155) to use `:root[class~="dark"]` instead of `.dark`. The problem was that the "light" theme wasn't looking as it should. Regardless of my mac's settings (dark/light didn't affect the issue) I tried the following:

Using "Arc browser" from `menubar > View > Appearance` I tried to switch the browser's appearance and see how it affects the theme. BTW I used the [`useTheme`](https://github.com/pacocoursey/next-themes?tab=readme-ov-file#usetheme) hook to set my theme and the results were the following regardless mac's theme:

* when the browser's appearance was set to light theme, switching was working correctly
* when the browser's appearance was set to dark theme, light theme had issues

Trying your recommendation to replace `@media (prefers-color-scheme: dark)` with `:root[class~="dark"]` seems to fixes the issue.

**Comment by rhyek** [comment]
@chrispetrou could you instead try removing this code from the top of your globals.css if you have it?

```css
:root {
  --background: #ffffff;
  --foreground: #171717;
}

@media (prefers-color-scheme: dark) {
  --background: #0a0a0a;
  --foreground: #ededed;
}
```

I think these are not needed and in fact I was seeing some inconsistencies, still. Some parts of shadcn components use the css `hsl` function with the variables defined in global.css, but those I copied here are not valid hsl values.

**Comment by chrispetrou** [comment]
@rhyek just tried it, removing them seems to have no effect visually. 
In a way it makes some sense for them to be redundant since the `@layer base` directive from Tailwind is more specific and generally loaded after other styles, which means it can override earlier root styles.
So by having the following:
```
@layer base {
  :root {
    --background: 0 0% 100%;
    ...
```
its ensured that the styles defined in `@layer base` are applied, regardless of what was previously defined

**Comment by alxgrgrs** [comment]
On Arc Browser specifically, there seems to be a continuous issue with the dark mode colors. From what it looks like, there seems that arc is trying to create its own Dark mode color scheme. Chrome plays nice with dark mode. 

Neither Chrome or Arc have any plugins, total base browser installations.

**Comment by chrispetrou** [comment]
@alxgrgrs using the recommendation from the [comment here](https://github.com/shadcn-ui/ui/issues/313#issuecomment-2336463000) seems to fix the issue kind of...
The thing is that when the user has chosen to use e.g. `dark` theme on Arc but `light` theme on mac OS using the following snippet:
```ts
const { theme, setTheme } = useTheme()
...
<button onClick={() => setTheme('system')}>System Mode</button>
```
will result in interpreting as `system` preference the `dark` theme, so there must be a conflict between the OS and the browser regarding of what is considered as "system" theme-preference. I don't think this is a direct `next-themes` issue though, or even `shadcn-ui` related.

Although what [this comment](https://github.com/shadcn-ui/ui/issues/313#issuecomment-2336463000) addresses and [my comment](https://github.com/shadcn-ui/ui/issues/313#issuecomment-2353788595) confirms is that without replacing `@media (prefers-color-scheme: dark)` with `:root[class~="dark"]` there seems to be a light theme glitch regardless of which browser you're using. Not sure if others have the same issue though...

**Comment by li-kang** [comment]
This fix it for me. use `:root.dark-mode` replace `.dark`

```js
@tailwind base;
@tailwind components;
@tailwind utilities;
 
@layer base {
  :root {
    --background: 0 0% 100%;
    --foreground: 222.2 84% 4.9%;
 
    --muted: 210 40% 96.1%;
    --muted-foreground: 215.4 16.3% 46.9%;
 
    --popover: 0 0% 100%;
    --popover-foreground: 222.2 84% 4.9%;
 
    --card: 0 0% 100%;
    --card-foreground: 222.2 84% 4.9%;
 
    --border: 214.3 31.8% 91.4%;
    --input: 214.3 31.8% 91.4%;
 
    --primary: 222.2 47.4% 11.2%;
    --primary-foreground: 210 40% 98%;
 
    --secondary: 210 40% 96.1%;
    --secondary-foreground: 222.2 47.4% 11.2%;
 
    --accent: 210 40% 96.1%;
    --accent-foreground: 222.2 47.4% 11.2%;
 
    --destructive: 0 84.2% 60.2%;
    --destructive-foreground: 210 40% 98%;
 
    --ring: 222.2 84% 4.9%;
 
    --radius: 0.5rem;
  }
 
  :root.dark-mode {
    --background: 222.2 84% 4.9%;
    --foreground: 210 40% 98%;
 
    --muted: 217.2 32.6% 17.5%;
    --muted-foreground: 215 20.2% 65.1%;
 
    --popover: 222.2 84% 4.9%;
    --popover-foreground: 210 40% 98%;
 
    --card: 222.2 84% 4.9%;
    --card-foreground: 210 40% 98%;
 
    --border: 217.2 32.6% 17.5%;
    --input: 217.2 32.6% 17.5%;
 
    --primary: 210 40% 98%;
    --primary-foreground: 222.2 47.4% 11.2%;
 
    --secondary: 217.2 32.6% 17.5%;
    --secondary-foreground: 210 40% 98%;
 
    --accent: 217.2 32.6% 17.5%;
    --accent-foreground: 210 40% 98%;
 
    --destructive: 0 62.8% 30.6%;
    --destructive-foreground: 210 40% 98%;
 
    --ring: 212.7 26.8% 83.9%;
  }
}
 
@layer base {
  * {
    @apply border-border;
  }
  body {
    @apply bg-background text-foreground;
  }
}
```

**Comment by musjj** [comment]
@Zwyx 

Thank you, it drove me crazy why it was working on *some* of my projects! To be more precise, you don't need to use `dark:` specifically. Tailwind's scanner is very coarse, so having the word `dark` *anywhere* in your codebase is enough for it to be detected (see https://tailwindcss.com/docs/content-configuration#class-detection-in-depth). My problematic codebase didn't include that word so it didn't work.

So if you want to guarantee that the classname never gets dropped, you should always safelist `dark`. This should be included in the default config IMO.

**Comment by Zwyx** [comment]
Ah yes that's interesting, having `dark` anywhere would work... Thank you for this detail.

**Comment by Ekefrancisokechukwu** [comment]
> Had the same issue, fixed it with `:root[class~="dark"]` instead of `.dark` in `style.css`
> 
> style.css for reference: https://github.com/langfuse/langfuse-docs/blob/e9ba748e0e53e5d9179ffc0040839108d1930621/style.css#L39

thanks this worked for me

**Comment by ethanenkhtur** [comment]
Replacing `.dark` with specifically `:root[data-theme='dark']` was the only way it worked for me, guys!
